src/components/permissions/rowLevelPolicyHelpers.ts

```diff
@@ -0,0 +1,332 @@
+  // If using clause is not used return nothing
+  // otherwise return USING from usingCondition (view rule, see compileRowLevelPolicies) or condition (current rule condition)
+  const getUsingClause = () => {
+    if (!ruleSettings.using) return ''
+    const conditionToUse = ruleSettings.withCheck ? usingCondition : condition
+    return `USING ${addBracketsAndPermissionCheck(conditionToUse)}`
+  }
+
+  const ruleSettings = ruleTypes[ruleType]
+
+  if (!ruleSettings) {
+    console.warn(`rule ${ruleType} does not exist, ignoring`)
+    return ''
+  }
+
+  const create = `CREATE POLICY "${ruleType}_${permissionAbbreviation}"`
+  const onFor = `ON "${tableName}" FOR ${ruleSettings.for}`
+
+  const withCheck = ruleSettings.withCheck
+    ? `WITH CHECK ${addBracketsAndPermissionCheck(condition)}`
+    : ''
+
+  // Combine them all
+  return `${create} ${onFor} ${getUsingClause()} ${withCheck}`
+}
+
+// Creates SQL from JSON, as per https://github.com/planetarydev/json-sql-builder2, but only for WHERE clause content
+// in { job_title: { $in: ['Sales Manager', 'Account Manager'] }, country_code: 'RU' }
+// out `job_title IN ('Sales Manager', 'Account Manager') AND country_code = 'RU'`
+const getSqlConditionFromJSON = (jsonSQLcondition: any) => {
+  // with no select field specified will default to SELECT *
+  const { sql, values } = jsonToSql.$select({
+    $where: {
+      ...jsonSQLcondition,
+    },
+  })
+
+  let query = sql
+  // $select return key values, which need to be replaced
+  // count backwards in case ${x} above 9 exists
+  for (let i = values.length - 1; i >= 0; i--) {
+    query = query.replace(`$${i + 1}`, values[i])
+  }
+  // Finally return only the WHERE part
+  return query.replace('SELECT * WHERE ', '')
+}
+
+// Replaces 'placeholder' templated values in a string
+// in `hello there, in jwt there is this user id: jwtUserDetails_bigint_userId, and this is permission specific template_id: jwtPermission_bigint_templateId`, `somePermissionAbbreviation`
+// out `hello there, in jwt there is this user id: jwt_get_bigint('userId'), and this is permission specific template_id: jwt_get_bigint('somePermissionAbbreviation_tempalteId')`
+const replacePlaceholders = (sql: string, permissionAbbreviation: string) => {
+  let resultSql = sql
+
+  const replacements = [
+    {
+      prefix: 'jwtUserDetails_bigint_',
+      prefixReplacement: "jwt_get_bigint('",
+      postfix: "')",
+    },
+    {
+      prefix: 'jwtPermission_bigint_',
+      prefixReplacement: `jwt_get_bigint('${permissionAbbreviation}_`,
+      postfix: "')",
+    },
+  ]
+
+  replacements.forEach(({ prefix, prefixReplacement, postfix }) => {
+    let match
+    const regex = new RegExp(`${prefix}\\w*`, 'g')
+    while ((match = regex.exec(resultSql))) {
+      const matchedValue = match[0]
+      const newValue = `${matchedValue.replace(prefix, prefixReplacement)}${postfix}`
+      resultSql = resultSql.replace(matchedValue, newValue)
+    }
+  })
+  return resultSql
+}
```

Comment:
This one also looks like it could be in a separated file. Just keep the core stuff in this one. Then it might be easier to follow. I'm still trying and it's quite hard.

Reply:
Sure can move it, once we agree on a good name for that utility file.

Btw, just trying to understand, you've made a comment

 > but since it contains everything is quite hard to follow the line

Is it the length of the file that's making it hard to follow, or excessive break down of logic into methods, so you need to jump from one to another (or maybe the method naming) ?

If it's the "excessive break down of logic into methods, so you need to jump from one to another", then would putting some of the methods in another file make it easier or more difficult ?

Reply:
I've move getSqlConditionFromJSON to helpersUtils: dfb94899f966a33a7b50c39a337dcc9869732288

But I've kept `replacePlaceholders` here, since I thin it's part of the core

Reply:
> I was thinking why I have some difficult on following the code, you have heaps of comments (which is good in a sense, but would be easier if just reading the code was enough to understand what was going on)

Most of the comment describe input and output of methods, which is necessary in this case to understand what the code is trying to do. I personally not a fan of excessive comment, I think this level of complexity warrant more then just variable/method naming to explain functionality ( and it's one of those cases where it's hard to describe functionality in a few words that can be used to name variables/methods )

>  With so many things in the same file makes it hard to see nested and not-nested functions. 

Hmm, I think the suggestion you've made to move things to say utils methods, make it a bit easier 

> The order of how calls are made seems a bit random to me, but I haven't tried to debug to see, maybe it was just my impression.

it's mainly top down for core functional areas, (i.e. method will call within itself a method declared below), but not for helpers).